Lib/fontTools/designspaceLib/__init__.py

```diff
@@ -321,6 +665,117 @@ def setStyleMapFamilyName(self, styleMapFamilyName, languageCode="en"):
+                self.designLocation = {}
+            if axisName in self.designLocation:
+                del self.designLocation[axisName]
+            if self.userLocation is None:
+                self.userLocation = {}
+            if axisName in self.userLocation:
+                del self.userLocation[axisName]
+
+    def getLocationLabelDescriptor(self, doc: 'DesignSpaceDocument') -> Optional[LocationLabelDescriptor]:
+        """Get the :class:`LocationLabelDescriptor` instance that matches
+        this instances's :attr:`locationLabel`.
+
+        Raises if the named label can't be found.
+
+        .. versionadded:: 5.0
+        """
+        if self.locationLabel is None:
+            return None
+        label = doc.getLocationLabel(self.locationLabel)
+        if label is None:
+            raise DesignSpaceDocumentError(
+                'InstanceDescriptor.getLocationLabelDescriptor(): '
+                f'unknown location label `{self.locationLabel}` in instance `{self.name}`.'
+            )
+        return label
+
+    def getFullDesignLocation(self, doc: 'DesignSpaceDocument') -> AnisotropicLocationDict:
+        """Get the complete design location of this instance, by combining data
+        from the various location fields, default axis values and mappings, and
+        top-level location labels.
+
+        The source of truth for this instance's location is determined for each
+        axis independently by taking the first not-None field in this list:
+
+        - ``locationLabel``: the location along this axis is the same as the
+          matching STAT format 4 label. No anisotropy.
+        - ``designLocation[axisName]``: the explicit design location along this
+          axis, possibly anisotropic.
+        - ``userLocation[axisName]``: the explicit user location along this
+          axis. No anisotropy.
+        - ``axis.default``: default axis value. No anisotropy.
+
+        .. versionadded:: 5.0
+        """
+        label = self.getLocationLabelDescriptor(doc)
+        if label is not None:
+            return doc.map_forward(label.userLocation)  # type: ignore
+        result: AnisotropicLocationDict = {}
+        for axis in doc.axes:
+            if axis.name in self.designLocation:
+                result[axis.name] = self.designLocation[axis.name]
+            elif axis.name in self.userLocation:
+                result[axis.name] = axis.map_forward(self.userLocation[axis.name])
+            else:
+                result[axis.name] = axis.map_forward(axis.default)
+        return result
+
+    def getFullUserLocation(self, doc: 'DesignSpaceDocument') -> SimpleLocationDict:
+        """Get the complete user location for this instance.
+
+        .. seealso:: :meth:`getFullDesignLocation`
+
+        .. versionadded:: 5.0
+        """
+        return doc.map_backward(self.getFullDesignLocation(doc))
+
+    def getStatNames(self: InstanceDescriptor, doc: DesignSpaceDocument) -> StatNames:
```

Comment:
@anthrotype I'll also remove that method if I move the `statNames` module to `varLib`. However I'm using the STAT name generation stuff in the `convert5to4` method, so I'll have to depend on varLib there in order to keep calling the STAT name stuff. 

(that's because in DSv4 you have to write out the instance names yourself, but in v5 you can write nothing on the instances and it will all get generated from the STAT data)